In [1]:
import pandas as pd
import numpy as np
import re
from IPython.display import clear_output

In [2]:
df = pd.read_json('speeches.json',orient='records')

In [3]:
df.iloc[0].to_dict()['segments']

{'segment': 0,
 'tag': 'RESOCONTO',
 'start': 'RESOCONTO STENOGRAFICO',
 'president': 0.128916308276516,
 'page': 0.032191545690914,
 'len': 2,
 'score': 1.484956310433453,
 'year': 2015,
 'month': 6,
 'day': 8,
 'leg': 17,
 'id': 370,
 'convocation': 25231,
 'date': '2015-06-08 00:00:00',
 'title': 'Seduta di lunedì 8 giugno 2015',
 'text': 'RESOCONTO STENOGRAFICO 437. ',
 'presidenza': False,
 'speech': False}

{"segment": 0, "tag": "RESOCONTO", "start": "RESOCONTO STENOGRAFICO", "president": 0.12891630827651634, "page": 0.032191545690914675, "len": 2, "score": 1.4849563104334538, "year": 2015, "month": 6, "day": 8, "leg": 17, "id": 370, "convocation": 25231, "date": "2015-06-08 00:00:00", "title": "Seduta di luned\u00ec 8 giugno 2015", "text": "RESOCONTO STENOGRAFICO 437. ", "presidenza": false, "speech": false},

In [4]:
data = []
l = len(df.index)
for index in range(l):
    info = df.iloc[index].to_dict()['segments']
    data.append(info)
    if index%1000==0:
        clear_output(wait=True)
        print('done '+str(index)+'/'+str(l)+'    '+str((index/l)*100)+'%')

clear_output(wait=False)
text = pd.DataFrame.from_dict(data)
text = text[text['score']>2.5]

In [5]:
r = text.iloc[0]
id=r['id']
t=r['text']
surname=r['surname']
name=r['name']
fname=surname+'_'+name+'_'+str(id)+'.txt'
print(fname)
print(t)

MARTI_ROBERTO_370.txt
ROBERTO GIACHETTI La seduta comincia alle 14,05. 


In [ ]:
def CreateCorpusFromDataFrame(corpusfolder,dataf):
    for index, r in dataf.iterrows():
        id=r['id']
        t=r['text'].upper().replace("À","A'").replace("È","E'").replace("É","E'").replace("Ì","I'").replace("Í","I'").replace("Ò","O'").replace("Ó","O'").replace("Ù","U'").replace("Ú","U'").replace("Ü","U'")
        surname=r['surname']
        name=r['name']
        fname=surname+' '+name+'_'+str(id)+'.txt'
        corpusfile=open(corpusfolder+'/'+fname,'a', encoding="utf-8")
        corpusfile.write(t.lower())
        corpusfile.close()
        clear_output(wait=True)
        print('done '+str(index)+'/'+str(len(dataf.index))+'    '+str((index/len(dataf.index))*100)+'%')
    clear_output(wait=True)
    print('done')
    return None

CreateCorpusFromDataFrame('./Corpus',text)

done 333589/1197023    27.868219741809476%


In [19]:
text.iloc[339]['text']

'MARGHERITA BONIVER, Sottosegretario di Stato per gli affari esteri, concorda. La Camera, con votazioni nominali elettroniche, approva l’emendamento 19.1 (ex articolo 86, comma 4-bis, del regolamento), l’articolo 19, nel testo emendato, nonché gli articoli 20, 21 e 22, ai quali non sono riferiti emendamenti. '

In [2]:
from nltk.corpus.reader import CategorizedPlaintextCorpusReader
my_corpus=CategorizedPlaintextCorpusReader('./Corpus/', r'.*', cat_pattern=r'(.*)_.*') 

In [3]:
my_corpus.categories()

['ABATERUSSO_ERNESTO',
 'ABATE_COSIMO',
 'ABBATANGELO_MASSIMO',
 'ABBATE_FABRIZIO',
 'ABBATE_MICHELE',
 'ABBIATI_AMAELE',
 'ABBIATI_DOLORES',
 'ABBONDANZIERI_MARISA',
 'ABBRUZZESE_GIUSEPPE',
 'ABBRUZZESE_SALVATORE',
 'ABELLI_GIAN CARLO',
 'ABELLI_TULLIO',
 'ABENANTE_ANGELO',
 'ABETE_GIANCARLO',
 'ABRIGNANI_IGNAZIO',
 'ACCAME_FALCO',
 'ACCIARINI_MARIA CHIARA',
 'ACCIARO_GIOVANNI CARLO',
 'ACCREMAN_VENIERO',
 'ACERBO_MAURIZIO',
 'ACHILLI_MICHELE',
 'ACIERNO_ALBERTO',
 'ACQUAROLI_FRANCESCO',
 'ACQUARONE_LORENZO',
 'ACUNZO_NICOLA',
 'ADAMOLI_GELASIO',
 'ADAMO_NICOLA',
 'ADDUCE_SALVATORE',
 'ADELIZZI_COSIMO',
 'ADENTI_FRANCESCO',
 'ADINOLFI_MARIO',
 'ADOLFO_VITTORIO',
 'ADONNINO_GIOVAN BATTISTA',
 'ADORNATO_FERDINANDO',
 'AFFRONTI_PAOLO',
 'AGLIETTA_MARIA ADELAIDE',
 'AGNALETTI_ANDREA',
 'AGNELLI_SUSANNA',
 'AGOSTA_MATTEO',
 'AGOSTINACCHIO_PAOLO ANTONIO MARIO',
 'AGOSTINELLI_DONATELLA',
 'AGOSTINI_LUCIANO',
 'AGOSTINI_MAURO',
 'AGOSTINI_ROBERTA',
 'AGRIMI_ALESSANDRO',
 'AGRUSTI_MICHELANGELO

In [4]:
my_corpus.words(categories='ABATE_COSIMO')

['abate', '.', 'quelli', 'che', 'avete', 'sfruttato', ...]

In [7]:
my_corpus.sents()

[['abaterusso', 'ernesto', 'f', 'f', 'f', 'f', 'f', 'f', 'c', 'c', 'f', 'c', 'c', 'f', 'c', 'f', 'f', 'f', 'f', 'f', 'c', 'f', 'c', 'c', 'c', 'f', 'c', 'c', 'f', 'f', 'f', 'f', 'f', 'f', 'f', 'f'], ['abaterusso', 'ernesto', '1', 'f', 'f', 'f', 'f', 'f', 'c', 'f', 'c', 'p'], ...]

In [6]:
my_corpus.raw()

KeyboardInterrupt: 

In [5]:
from sklearn.datasets import fetch_20newsgroups
from sklearn.feature_extraction.text import TfidfVectorizer
import nltk
from nltk.corpus import stopwords
from string import punctuation

In [6]:
stops = stopwords.words('italian')

In [7]:
import spacy
nlp = spacy.load("it_core_news_lg")
lemmatizer = nlp.get_pipe("lemmatizer")

In [8]:
doc = nlp("Ieri sono andato in due supermercati. Oggi volevo andare all'ippodromo. Stasera mangio la pizza con le verdure.")
print([token.lemma_ for token in doc])

['ieri', 'essere', 'andare', 'in', 'due', 'supermercato', '.', 'oggi', 'volere', 'andare', 'a il', 'ippodromo', '.', 'stasera', 'mangiare', 'il', 'pizza', 'con', 'il', 'verdura', '.']


In [9]:
def tokenize(text):
    res = []
    doc = nlp(text)
    for token in doc:
        if token.text not in punctuation and token.text not in stops:
            res.append(token)
    return res

In [10]:
tokenize("Ieri sono andato in due supermercati. Oggi volevo andare all'ippodromo. Stasera mangio la pizza con le verdure.")

[Ieri,
 andato,
 due,
 supermercati,
 Oggi,
 volevo,
 andare,
 all',
 ippodromo,
 Stasera,
 mangio,
 pizza,
 verdure]

In [11]:
tokenize(text.iloc[0]['text'])

[ROBERTO, GIACHETTI, La, seduta, comincia, 14,05]

In [ ]:
def doc_vocabulary(doc_id, model, vocabulary, top_w=10):
    words_scores = sorted(enumerate(model[doc_id]), key=lambda x: -x[1])
    return [(vocabulary[wid], score) for wid, score in words_scores[:top_w]]

In [ ]:
doc = nlp(text.iloc[0]['text'])
for token in doc:
    print(token.text, token.has_vector, token.vector_norm, token.is_oov) #token.vector

In [ ]:
text.iloc[0]['persona'].split('p')[-1]

In [12]:
# load info about deputy
df1 = pd.read_csv('parlamentari0x.csv')
df2 = pd.read_csv('parlamentari1x.csv')
dfp = pd.concat([df1, df2], ignore_index=True)
# transform date from int to date format
dfp['inizioMandato'] = pd.to_datetime(dfp['inizioMandato'], format="%Y%m%d")
dfp['fineMandato'] = dfp['fineMandato'].fillna(20990101)  # if person is currently active use placeholder as end date 
dfp['fineMandato'] = pd.to_datetime(dfp['fineMandato'], format="%Y%m%d")
# loan info about senators
df1 = pd.read_csv('senato0x.csv')
df2 = pd.read_csv('senato1x.csv')
dfs = pd.concat([df1, df2], ignore_index=True)
dfs['inizioMandato'] = pd.to_datetime(dfs['inizioMandato'], format="%Y%m%d")
dfs['fineMandato'] = dfs['fineMandato'].fillna(20990101)  # if person is currently active use placeholder as end date 
dfs['fineMandato'] = pd.to_datetime(dfs['fineMandato'], format="%Y%m%d")
# load info about ministers
df1 = pd.read_csv('ministri0x.csv')
df2 = pd.read_csv('ministri1x.csv')
dfm = pd.concat([df1, df2], ignore_index=True)
# rename some colums (to be abble to concat)
dfm = dfm.rename(columns={"d": "persona"})
dfp = dfp.rename(columns={"inizioMandato": "dataInizio", "fineMandato": "dataFine"})
dfs = dfs.rename(columns={"inizioMandato": "dataInizio", "fineMandato": "dataFine"})
# replace ministers link to it's personal link (have same numerical number, need only to change prefix and remove '_xx')
for i, d in dfm.iterrows():
    m = re.search('deputato.rdf/d(.+?)_', d['persona'])
    dfm.at[i,'persona'] = 'http://dati.camera.it/ocd/persona.rdf/p'+m.group(1)
# transform date from int to date format   
dfm['dataInizio'] = pd.to_datetime(dfm['dataInizio'], format="%Y%m%d")
dfm['dataFine'] = dfm['dataFine'].fillna(20990101)
dfm['dataFine'] = pd.to_datetime(dfm['dataFine'], format="%Y%m%d")
# need only some colums
dfp = dfp[['persona','cognome','nome','dataInizio','dataFine']]
dfm = dfm[['persona','cognome','nome','dataInizio','dataFine','carica','nomeOrganoGoverno']]
# final concat into one df
filtered_personnel = pd.concat([dfp, dfm, dfs], ignore_index=True)
filtered_personnel = filtered_personnel.rename(columns={"cognome": "surname", "nome": "name"})
filtered_personnel =  filtered_personnel[['persona','surname','name','dataInizio','dataFine','carica','nomeOrganoGoverno']]

In [26]:
i = 0
date = filtered_personnel[(filtered_personnel["dataInizio"] <= text.iloc[i]['date']) & (filtered_personnel["dataFine"] >= text.iloc[i]['date'])]
tokens = tokenize(text.iloc[i]['text'])
simmilarity = []
j = 0
for index, row in date.iterrows():
    sur_token = nlp(row['surname'])
    name_token = nlp(row['name'])
    sur = None
    name = None
    sur_sim = 0
    name_sim = 0
    for token in tokens:
        if sur_token.similarity(token) > sur_sim:
            sur_sim = sur_token.similarity(token)
            sur = row
        if name_token.similarity(token) > name_sim:
            name_sim = name_token.similarity(token)
            name = row
    simmilarity.append({'sur': sur, 'sur_sim': sur_sim, 'name':name, 'name_sim':name_sim})
    clear_output(wait=True)
    j +=1
    print('done '+str(j)+'/'+str(len(date.index))+'    '+str((j/len(date.index))*100)+'%')
simmilarity = pd.DataFrame.from_dict(simmilarity)

done 982/982    100.0%


In [24]:
sur_token = nlp(filtered_personnel.iloc[7435]['surname'])
name_token = nlp(filtered_personnel.iloc[7435]['name'])
sur = None
name = None
sur_sim = 0
name_sim = 0
for token in tokens:
    if sur_token.similarity(token) > sur_sim:
        sur_sim = sur_token.similarity(token)
        sur = row
    if name_token.similarity(token) > name_sim:
        name_sim = name_token.similarity(token)
        name = row
print(sur_sim, name_sim)

1.0 1.0


In [ ]:
text.iloc[0]['text']

'ROBERTO GIACHETTI La seduta comincia alle 14,05. '

In [21]:
date.loc[filtered_personnel['surname'] == "GIACHETTI"]

persona    surname     name  \
7435  http://dati.camera.it/ocd/persona.rdf/p300480  GIACHETTI  ROBERTO   

     dataInizio   dataFine carica nomeOrganoGoverno  
7435 2013-03-05 2018-03-22    NaN               NaN

In [27]:
simmilarity

sur   sur_sim  \
0                                                 None  0.000000   
1    persona              http://dati.camera.it/ocd...  0.504736   
2    persona              http://dati.camera.it/ocd...  0.615276   
3    persona              http://dati.camera.it/ocd...  0.606194   
4                                                 None  0.000000   
..                                                 ...       ...   
977                                               None  0.000000   
978                                               None  0.000000   
979                                               None  0.000000   
980                                               None  0.000000   
981                                               None  0.000000   

                                                  name  name_sim  
0    persona              http://dati.camera.it/ocd...  0.762184  
1    persona              http://dati.camera.it/ocd...  0.836266  
2    persona              http://dati.camera.it/ocd...  0.742634  
3    persona              http://dati.camera.it/ocd...  0.556078  
4    persona              http://dati.camera.it/ocd...  0.796587  
..                                                 ...       ...  
977  persona              http://dati.camera.it/ocd...  0.826297  
978  persona              http://dati.camera.it/ocd...  0.605342  
979  persona              http://dati.camera.it/ocd...  0.711881  
980  persona              http://dati.camera.it/ocd...  0.750619  
981  persona              http://dati.camera.it/ocd...  0.823274  

[982 rows x 4 columns]

In [31]:
simmilarity[(simmilarity['sur_sim'] == simmilarity['sur_sim'].max()) | (simmilarity['name_sim'] == simmilarity['name_sim'].max())]

sur   sur_sim  \
25   persona              http://dati.camera.it/ocd...  1.000000   
120                                               None  0.000000   
256                                               None  0.000000   
378  persona              http://dati.camera.it/ocd...  0.205963   
414                                               None  0.000000   
417  persona              http://dati.camera.it/ocd...  0.402618   
460  persona              http://dati.camera.it/ocd...  0.578858   
527                                               None  0.000000   
621                                               None  0.000000   
623  persona              http://dati.camera.it/ocd...  0.817134   
675                                               None  0.000000   
680                                               None  0.000000   
821  persona              http://dati.camera.it/ocd...  0.682343   
957  persona              http://dati.camera.it/ocd...  0.654393   

                                                  name  name_sim  
25   persona              http://dati.camera.it/ocd...       1.0  
120  persona              http://dati.camera.it/ocd...       1.0  
256  persona              http://dati.camera.it/ocd...       1.0  
378  persona              http://dati.camera.it/ocd...       1.0  
414  persona              http://dati.camera.it/ocd...       1.0  
417  persona              http://dati.camera.it/ocd...       1.0  
460  persona              http://dati.camera.it/ocd...       1.0  
527  persona              http://dati.camera.it/ocd...       1.0  
621  persona              http://dati.camera.it/ocd...       1.0  
623  persona              http://dati.camera.it/ocd...       1.0  
675  persona              http://dati.camera.it/ocd...       1.0  
680  persona              http://dati.camera.it/ocd...       1.0  
821  persona              http://dati.camera.it/ocd...       1.0  
957  persona              http://dati.camera.it/ocd...       1.0

In [13]:
text2 = text[['text','persona','date','surname','name']].sample(100)
text2 = text2.reset_index(drop=True)

In [41]:
def find(t, date, personell):
    regex = re.compile("[^a-zA-Z -']")
    t = regex.sub('', t)
    data = personell[(personell["dataInizio"] <= date) & (personell["dataFine"] >= date)]
    tokens = tokenize(t)
    simmilarity = []
    doc = nlp(t)
    for index, row in data.iterrows():
        sur_token = nlp(row['surname'])
        name_token = nlp(row['name'])
        person = row['persona']
        sur_sim = 0
        name_sim = 0

        for token in tokens:
            if sur_token.similarity(token) > sur_sim:
                sur_sim = sur_token.similarity(token)
            if name_token.similarity(token) > name_sim:
                name_sim = name_token.similarity(token)
        simmilarity.append({'sur': row['surname'], 'sur_sim': sur_sim, 'name':row['name'], 'name_sim':name_sim, 'person':person})
    simmilarity = pd.DataFrame.from_dict(simmilarity)
    simmilarity = simmilarity[(simmilarity['sur_sim'] == simmilarity['sur_sim'].max()) | (simmilarity['name_sim'] == simmilarity['name_sim'].max())]
    if len(simmilarity.index) > 1:
        if 1.0 in simmilarity['sur_sim']:
            if len(simmilarity[(simmilarity['sur_sim'] == simmilarity['sur_sim'].max())].index) > 1:
                return simmilarity[(simmilarity['sur_sim'] == simmilarity['sur_sim'].max()) & (simmilarity['name_sim'] == simmilarity['name_sim'].max())].iloc[0]
            else:
                return simmilarity[(simmilarity['sur_sim'] == simmilarity['sur_sim'].max())].iloc[0]
        else:
            if 1.0 in simmilarity['name_sim']:
                sim = simmilarity[simmilarity['name_sim'] == 1.0]
                return sim[(sim['sur_sim'] == sim['sur_sim'].max())].iloc[0]
            else:
                simmilarity['score'] = simmilarity['sur_sim']*1.5 + simmilarity['name_sim']
                return simmilarity[(simmilarity['score'] == simmilarity['score'].max())].iloc[0]
    else: 
        return simmilarity.iloc[0]

In [15]:
l = len(text2.index)
res = []

for index, row in text2.iterrows():
    person = find(row['text'], row['date'], filtered_personnel)['person']
    predicted = filtered_personnel.loc[filtered_personnel['persona'] == person]
    if person is not None:
        pred_surname = predicted.iloc[0]['surname']
        pred_name = predicted.iloc[0]['name']
    else:
        pred_surname = None
        pred_name = None
    if row["persona"] == person:
        correct = True
    else:
        correct = False
    
    res.append({"True persona": row["persona"], "True name": row["name"], "True surname": row["surname"],
                "Text":row["text"],  "Correct":correct, "Predicted persona": person,
                "Predicted surname": pred_surname, 
                "Predicted name": pred_name})
    
    clear_output(wait=True)
    print('done '+str(index)+'/'+str(l)+'    '+str((index/l)*100)+'%')
            
clear_output(wait=True)
print('done')
result = pd.DataFrame.from_dict(res)
result['Correct'].describe()

done


count      100
unique       2
top       True
freq        73
Name: Correct, dtype: object

In [16]:
pd.set_option("display.max_rows", None)
result[result['Correct'] == False]

True persona            True name  \
2   http://dati.camera.it/ocd/persona.rdf/p300330             GIUSEPPE   
6     http://dati.camera.it/ocd/persona.rdf/p7610                MARIO   
15   http://dati.camera.it/ocd/persona.rdf/p23060                MARIO   
20   http://dati.camera.it/ocd/persona.rdf/p35870  MARIELLA MARIA RITA   
22   http://dati.camera.it/ocd/persona.rdf/p19550             GIUSEPPE   
30   http://dati.camera.it/ocd/persona.rdf/p13240               PIETRO   
31  http://dati.camera.it/ocd/persona.rdf/p307278              ANTONIO   
38   http://dati.camera.it/ocd/persona.rdf/p12810             VINCENZO   
40  http://dati.camera.it/ocd/persona.rdf/p300486                 ENZO   
42   http://dati.camera.it/ocd/persona.rdf/p31040              ALBERTO   
44    http://dati.camera.it/ocd/persona.rdf/p2060             BERNARDO   
46   http://dati.camera.it/ocd/persona.rdf/p21830             GIROLAMO   
52   http://dati.camera.it/ocd/persona.rdf/p50137           ALESSANDRO   
58    http://dati.camera.it/ocd/persona.rdf/p3640                OSCAR   
60    http://dati.camera.it/ocd/persona.rdf/p2680               FRANCO   
63   http://dati.camera.it/ocd/persona.rdf/p19550             GIUSEPPE   
69   http://dati.camera.it/ocd/persona.rdf/p26850                MARIO   
72     http://dati.camera.it/ocd/persona.rdf/p950               ARRIGO   
76   http://dati.camera.it/ocd/persona.rdf/p24900             VINCENZO   
79    http://dati.camera.it/ocd/persona.rdf/p3040         MARIO MARINO   
81   http://dati.camera.it/ocd/persona.rdf/p29140            SALVATORE   
84   http://dati.camera.it/ocd/persona.rdf/p24180           ANNA MARIA   
86   http://dati.camera.it/ocd/persona.rdf/p22070               TOMASO   
89   http://dati.camera.it/ocd/persona.rdf/p50057           GIAN PAOLO   
92  http://dati.camera.it/ocd/persona.rdf/p303108                CARLO   
96    http://dati.camera.it/ocd/persona.rdf/p7320    GIOVANNI BATTISTA   
99   http://dati.camera.it/ocd/persona.rdf/p50236              ALFREDO   

                     True surname  \
2                           GALLO   
6                      COTELLESSA   
15                         CHELLA   
20                 CAVANNA SCIREA   
22                     PELLEGRINO   
30                        GERMANI   
31                       FEDERICO   
38                       SELVAGGI   
40                          RAISI   
42                       TRIDENTE   
44                       D'AREZZO   
46                          RALLO   
52                            CE'   
58                         MAMMI'   
60                        FRANCHI   
63                     PELLEGRINO   
69                         ROVEDA   
72                       BOLDRINI   
76                           VITI   
79                      GUADALUPI   
81                      CARDINALE   
84                       PEDRAZZI   
86  STAITI DI CUDDIA DELLE CHIUSE   
89                          LANDI   
92                          SARRO   
96                         CARRON   
99                      MANTOVANO   

                                                 Text  Correct  \
2   GIUSEPPE ROSSIELLO. No, signor Presidente, ins...    False   
6   (7600) (( ~COTELLESSA )). (( ‘11 sottoscritto ...    False   
15  MARIO POCHETTI. Chiediamo lo scrutinio segreto...    False   
20                     CAVANNA SCIREA MARIELLA V c F     False   
22  PELLEGRINO . A nostro giudizio, non può essere...    False   
30            iGERMANI, Relatore. Chiede di parlare.     False   
31  FEDERICO FORNARO (LEU). Non si può ovviamente ...    False   
38  SELVAGGI. E torniamo al bilancio. Se analizzia...    False   
40  ENZO MOAVERO MILANESI, Ministro per gli affari...    False   
42  PRESIDENTE . . 94339, 94340, 94341, 94342 , 94...    False   
44  D ' AREZZO . — Al Ministro dei lavori pubblici...    False   
46  RALLO, DEL DONNO, BAGHINO. Al terzo comma, all...    False   
52               ALESSANDRO CÈ. Non è nel merito !     False   
58 

In [39]:
regex = re.compile("[^a-zA-Z -']")
t = regex.sub('', text2.iloc[20]['text'])
token = tokenize(t)
sur_token = nlp('CAVANNA SCIREA')
sur_token.similarity(token[0])
sur_token.similarity(nlp(text2.iloc[20]['text']))

C:\Users\Anton\AppData\Local\Temp\ipykernel_14236\3388924094.py:5: UserWarning: [W008] Evaluating Doc.similarity based on empty vectors.
  sur_token.similarity(token[0])
C:\Users\Anton\AppData\Local\Temp\ipykernel_14236\3388924094.py:6: UserWarning: [W008] Evaluating Doc.similarity based on empty vectors.
  sur_token.similarity(nlp(text2.iloc[20]['text']))


0.0

In [40]:
from spacy.matcher import PhraseMatcher
matcher = PhraseMatcher(nlp.vocab)
terms = ["CAVANNA SCIREA", "MARIELLA MARIA RITA"]
patterns = [nlp.make_doc(text) for text in terms]
matcher.add("TerminologyList", patterns)
doc = nlp(text2.iloc[20]['text'])
matches = matcher(doc)
for match_id, start, end in matches:
    span = doc[start:end]
    print(span.text)

CAVANNA SCIREA


In [35]:
regex = re.compile("[^a-zA-Z -']")
personell = filtered_personnel
t = regex.sub('', text2.iloc[20]['text'])
data = personell[(personell["dataInizio"] <= text2.iloc[20]['date']) & (personell["dataFine"] >= text2.iloc[20]['date'])]
tokens = tokenize(t)
simmilarity = []
for index, row in data.iterrows():
    sur_token = nlp(row['surname'])
    name_token = nlp(row['name'])
    person = row['persona']
    sur_sim = 0
    name_sim = 0
    for token in tokens:
        if sur_token.similarity(token) > sur_sim:
            sur_sim = sur_token.similarity(token)
            sur = row
        if name_token.similarity(token) > name_sim:
            name_sim = name_token.similarity(token)
            name = row
    simmilarity.append({'sur': row['surname'], 'sur_sim': sur_sim, 'name':row['name'], 'name_sim':name_sim, 'person':person})
simmilarity = pd.DataFrame.from_dict(simmilarity)
simmilarity

C:\Users\Anton\AppData\Local\Temp\ipykernel_14236\4174560876.py:14: UserWarning: [W008] Evaluating Doc.similarity based on empty vectors.
  if sur_token.similarity(token) > sur_sim:
C:\Users\Anton\AppData\Local\Temp\ipykernel_14236\4174560876.py:17: UserWarning: [W008] Evaluating Doc.similarity based on empty vectors.
  if name_token.similarity(token) > name_sim:


sur   sur_sim                        name  name_sim  \
0                      ALOI  0.000000                   FORTUNATO  0.643837   
1                     COSTA  0.637772                    RAFFAELE  0.708381   
2                   CERULLO  0.000000                      PIETRO  0.727627   
3                  MASTELLA  0.783006              MARIO CLEMENTE  0.801801   
4                  OCCHETTO  0.000000                     ACHILLE  0.688819   
5                     SEGNI  0.377677  MARIOTTO GIOVANNI BATTISTA  0.767216   
6                BERLINGUER  0.445266                       LUIGI  0.710028   
7                 BASSANINI  0.000000                      FRANCO  0.787932   
8                    BIONDI  0.623957                     ALFREDO  0.701040   
9                     FIORI  0.520476                      PUBLIO  0.000000   
10                 MARTINAT  0.000000                         UGO  0.628155   
11                  PARLATO  0.399038                     ANTONIO  0.705225   
12                  SOSPIRI  0.000000                        NINO  0.706874   
13                    SPINI  0.000000                       VALDO  0.000000   
14                TATARELLA  0.000000                    GIUSEPPE  0.763083   
15                 VIOLANTE  0.570861                     LUCIANO  0.792419   
16              CRUCIANELLI  0.000000                     FAMIANO  0.000000   
17                 CORLEONE  0.607418                   FRANCESCO  0.722039   
18                CALDERISI  0.000000                    GIUSEPPE  0.763083   
19                 TRINGALI  0.000000                       PAOLO  0.716339   
20            AGOSTINACCHIO  0.000000         PAOLO ANTONIO MARIO  0.810237   
21                ANDREATTA  0.000000                   BENIAMINO  0.618205   
22                 BARZANTI  0.000000                        NEDO  0.000000   
23                   CASINI  0.530833             PIER FERDINANDO  0.730594   
24                  D'AIMMO  0.173787                    FLORINDO  0.000000   
25                     FINI  0.423272                  GIANFRANCO  0.715109   
26                 GUERZONI  0.000000                     LUCIANO  0.792419   
27               MATTARELLA  0.764584                      SERGIO  0.684223   
28                 MATTEOLI  0.000000                      ALTERO  0.000000   
29                  MAZZONE  0.661341                     ANTONIO  0.705225   
30             POLI BORTONE  0.698823                     ADRIANA  0.773861   
31                    VISCO  0.000000            VINCENZO ALFONSO  0.736395   
32                MONTECCHI  0.000000                       ELENA  0.637252   
33                 ANGELINI  0.703485                    GIORDANO  0.697106   
34                   ANGIUS  0.000000                      GAVINO  0.518956   
35                  BARGONE  0.000000                     ANTONIO  0.705225   
36                    BEEBE  0.000000                 CAROLE JANE  0.567524   
37                   BORDON  0.000000                      WILLER  0.000000   
38                   CAVERI  0.000000              LUCIANO EMILIO  0.764181   
39                  D'ALEMA  0.240238                     MASSIMO  0.662151   
40                   D'ALIA  0.245170                   SALVATORE  0.712659   
41                 DE JULIO  0.664319                      SERGIO  0.684223   
42              FINOCCHIARO  0.650564                        ANNA  0.810157   
43                FORMIGONI  0.000000                     ROBERTO  0.740433   
44        FUMAGALLI CARULLI  0.000000                  BATTISTINA  0.000000   
45                 GARAVINI  0.000000               ANDREA SERGIO  0.767315   
46                    LEONI  0.683459                    GIUSEPPE  0.763083   
47                      LIA  0.273147                     ANTONIO  0.705225   
48               LORENZETTI  0.000000                  MARIA RITA  0.804588   
49                   MASINI  0.573418                       NADIA  0.68820

In [40]:
import ipyparallel as ipp
# FIX THE CODE !!!

l = len(text2.index)
i = 0
res = []

def check_find(row, personnel):
    person = find(row['text'], row['date'], personnel)['person']
    predicted = personnel.loc[personnel['persona'] == person]
    if person is not None:
        pred_surname = predicted.iloc[0]['surname']
        pred_name = predicted.iloc[0]['name']
    else:
        pred_surname = None
        pred_name = None
    if row["persona"] == person:
        correct = True
    else:
        correct = False
    
    return {"True persona": row["persona"], "True name": row["name"], "True surname": row["surname"],
                "Text":row["text"],  "Correct":correct, "Predicted persona": person,
                "Predicted surname": pred_surname, 
                "Predicted name": pred_name}

def app_res(r):
    res.append(r)

with concurrent.futures.ProcessPoolExecutor() as executor:
    # for index, row in text2.iterrows():
    future_proc = {executor.submit(check_find, row, filtered_personnel): row for index, row in text2.iterrows()}
    for future in concurrent.futures.as_completed(future_proc):
        app_res(future.result())
        clear_output(wait=True)
        i += 1
        print('done '+str(i)+'/'+str(l)+'    '+str((i/l)*100)+'%')
result = pd.DataFrame.from_dict(res)
result['Correct'].describe()

BrokenProcessPool: A process in the process pool was terminated abruptly while the future was running or pending.